src/marqo/s2_inference/s2_inference.py

```diff
@@ -79,19 +93,39 @@ def _create_model_cache_key(model_name: str, device: str, model_properties: dict
 def _update_available_models(model_cache_key: str, model_name: str, validated_model_properties: dict,
                              device: str,
                              normalize_embeddings: bool) -> None:
-    """loads the model if it is not already loaded
+    """loads the model if it is not already loaded.
+    Note this method assume the model_properties are validated.
     """
+    model_size = get_model_size(model_name, validated_model_properties)
     if model_cache_key not in available_models:
-        try:
-            available_models[model_cache_key] = _load_model(model_name,
-                                                            validated_model_properties, device=device)
-            logger.info(f'loaded {model_name} on device {device} with normalization={normalize_embeddings}')
-        except:
-            raise ModelLoadError(
-                f"Unable to load model={model_name} on device={device} with normalization={normalize_embeddings}. "
-                f"If you are trying to load a custom model, "
-                f"please check that model_properties={validated_model_properties} is correct "
-                f"and the model has valid access permission. ")
+        if lock.locked():
+            raise ModelCacheManagementError("Request rejected, as this request attempted to update the model cache, while"
+                                            "another request was updating the model cache at the same time.\n"
+                                            "Please wait for 10 seconds and send the request again.\n"
+                                            "Marqo's documentation can be found here: `https://docs.marqo.ai/latest/`")
+        with lock:
+            _validate_model_into_device(model_name, validated_model_properties, device,
+                                       calling_func=_update_available_models.__name__)
+            try:
+                most_recently_used_time = datetime.datetime.now()
+                available_models[model_cache_key] = {AvailableModelsKey.model: _load_model(model_name,
+                                                                                           validated_model_properties,
+                                                                                           device=device,
+                                                                                           calling_func = _update_available_models.__name__),
+                                                     AvailableModelsKey.most_recently_used_time: most_recently_used_time,
+                                                     AvailableModelsKey.model_size: model_size}
+                logger.info(
+                    f'loaded {model_name} on device {device} with normalization={normalize_embeddings} at time={most_recently_used_time}.')
+            except:
```

Comment:
do we want to log the error that caused the ModelLoadError? Something like this:

logging.warning(f"Error loading model `{model_name}`, reason: {e} ")